### DISCUSSION TITLE
Evaluation step for textcat using transformer doesn't use GPU

### DISCUSSION THREAD

**Discussion by samehraban** [explosion/spaCy#10224] [Help: Other Questions]
## How to reproduce the behaviour
I noticed training a textcat using transformer is taking forever at the end of the first iteration. Turns out [this loop](https://github.com/explosion/spaCy/blob/0668a449ba68d7831626d9e2031cb84930db982c/spacy/language.py#L1415) takes about 40~50 seconds per iteration on my machine while the training itself takes about ~2.5 it/sec.

## Your Environment
- **spaCy version:** 3.2.1
- **Platform:** Linux-4.15.0-162-generic-x86_64-with-Ubuntu-18.04-bionic
- **Python version:** 3.6.12

**Comment by adrianeboyd** [comment]
The evaluation does use GPU if you're using GPU for training, but if you have a large dev corpus, it can take a noticeable amount of time. You might want to raise `eval_frequency` in your config or lower the size of your dev corpus (for a quick test, you can truncate it with `limit` in the `[corpora.dev]` block).

The batch sizes and other settings that manage the length of the training and eval steps can be set separately in the config, see: https://github.com/explosion/spaCy/discussions/8600#discussioncomment-969204

**Reply by samehraban** [reply]
I have about 4000 samples for dev corpus which I don't consider to be large. I looked at the GPU utilization during training and I realized that it's almost zero! Though it has occupied around 4.5 GB of GPU memory. This is not the case for textcat training without transformer.

**Reply by adrianeboyd** [reply]
Can you post your whole config? My first guess would be that the `textcat` isn't actually listening to the `transformer`? It looks like the quickstart does give you an unused `transformer` with `GPU`+`efficiency`, which is a bug.

To actually use a `transformer` with `textcat`, try `GPU`+`accuracy` instead?

**Reply by samehraban** [reply]
This is my whole transformer-base.cfg:

```
# This is an auto-generated partial config. To use it with 'spacy train'                                                                                                
# you can run spacy init fill-config to auto-fill all default settings:                                                                                                 
# python -m spacy init fill-config ./base_config.cfg ./config.cfg                                                                                                       
[paths]                                                                                                                                                                 
train = null                                                                                                                                                            
dev = null                                                                                                                                                              
                                                                                                                                                                        
[system]                                                                                                                                                                
gpu_allocator = "pytorch"                                                                                                                                               
                                                                                                                                                                        
[nlp]                                                                                                                                                                   
lang = "fa"                                                                                                                                                             
pipeline = ["transformer","textcat_multilabel"]                                                                                                                         
batch_size = 128                                                                                                                                                        
                                                                                                                                                                        
[components]                                                                                                                                                            
                                                                                                                                                                        
[components.transformer]                                                                                                                                                
factory = "transformer"                                                                                                                                                 
                                                                                                                                                                        
[components.transformer.model]                                                                                                                                          
@architectures = "spacy-transformers.TransformerModel.v3"                                                                                                               
name = "HooshvareLab/distilbert-fa-zwnj-base"                                                                                                                           
tokenizer_config = {"use_fast": true}                                                                                                                                   
                                                                                                                                                                        
[components.transformer.model.get_spans]                                                                                                                                
@span_getters = "spacy-transformers.strided_spans.v1"                                                                                                                   
window = 128                                                                                                                                                            
stride = 96                                                                                                                                                             
                                                                                                                                                                        
[components.textcat_multilabel]                                                                                                                                         
factory = "textcat_multilabel"                                                                                                                                          
                                                                                                                                                                        
[components.textcat_multilabel.model]                                                                                                                                   
@architectures = "spacy.TextCatEnsemble.v2"                                                                                                                             
nO = null                                                                                                                                                               
                                                                                                                                                                        
[components.textcat_multilabel.model.tok2vec]                                                                                                                           
@architectures = "spacy-transformers.TransformerListener.v1"                                                                                                            
grad_factor = 1.0
[components.textcat_multilabel.model.tok2vec.pooling]                                                                                                                   
@layers = "reduce_mean.v1"

[components.textcat_multilabel.model.linear_model]
@architectures = "spacy.TextCatBOW.v2"
exclusive_classes = false
ngram_size = 1
no_output_layer = false

[corpora]

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0

[training]
accumulate_gradient = 3
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
patience = 5000
max_steps = 30000
eval_frequency = 1000

[training.score_weights]
cats_macro_f = 1

[training.logger]                                                                                                                                                   
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = true

[training.optimizer]
@optimizers = "Adam.v1"

[training.optimizer.learn_rate]
@schedules = "warmup_linear.v1"
warmup_steps = 5000
total_steps = 20000
initial_rate = 5e-5

[training.batcher]
@batchers = "spacy.batch_by_padded.v1"
discard_oversize = true
size = 512
buffer = 256

[initialize]
vectors = ${paths.vectors}

**Reply by samehraban** [reply]
I just noticed that GPU utilization/train speed varies over time. I just watched it go on 5 it/sec for a couple of hundreds of steps and suddenly dropped to 2 sec/it. GPU utilization was about 40% and then dropped to almost zero.

**Reply by adrianeboyd** [reply]
The config looks fine, so my guess was not correct.

I think a lot of the BOW calculations are always on CPU, so it wouldn't be unexpected for the GPU utilization to vary more than with other components while training.

**Reply by samehraban** [reply]
But it drops to zero for evaluation. Why? And I'm not talking about some oscillation, It's dead zero.

**Reply by pmbaumgartner** [reply]
@samehraban - I was able to replicate what you saw. What's happening is that evaluation involves several steps that are entirely on CPU, e.g., reading in the dev docs (which are loaded from scratch every time to ensure a clean eval) and running the scorer. When you see the GPU use drop to 0, this is what's happening.

**Reply by samehraban** [reply]
@pmbaumgartner Actually I'm totally confused here. After about 2000 steps, GPU utilization drops to almost zero in training. If I set `discard_oversize` to `false` for the `minibatch_by_padded_size` it gets good but then after a while COOM occurs.

**Reply by pmbaumgartner** [reply]
In general, I think it's going to be difficult to replicate this without the same data you're using.

What are you using to profile/monitor your GPU and CPU?

**Reply by samehraban** [reply]
I use nvtop.

**Reply by pmbaumgartner** [reply]
If there's a script/repo/project that we can run to replicate the issue we can investigate a bit further, otherwise it's hard to know exactly what's going on in your second issue (OOM after 2000 steps).
